commands/importer/validate.go

```diff
@@ -0,0 +1,904 @@
+					FieldName:      fmt.Sprintf("direct_post.attachments[%d]", i),
+					Err:            fmt.Errorf("missing attachment file %q%s", attachmentPath, helpful),
+				}); err != nil {
+					return err
+				}
+			} else {
+				v.attachmentsUsed[attachmentPath]++
+			}
+		}
+	}
+
+	v.directPosts++
+
+	return nil
+}
+
+func (v *Validator) validateEmoji(info ImportFileInfo, line LineImportData) (err error) {
+	ive := validateNotNil(info, "emoji", line.Emoji, func(data EmojiImportData) *ImportValidationError {
+		appErr := validateEmojiImportData(&data)
+		if appErr != nil {
+			return &ImportValidationError{
+				ImportFileInfo: info,
+				FieldName:      "emoji",
+				Err:            appErr,
+			}
+		}
+
+		if data.Name != nil {
+			if existing, ok := v.emojis[*data.Name]; ok {
+				return &ImportValidationError{
+					ImportFileInfo: info,
+					FieldName:      "emoji",
+					Err:            fmt.Errorf("duplicate entry, previous was in line: %d", existing.CurrentLine),
+				}
+			}
+			v.emojis[*data.Name] = info
+		}
+
+		if !v.ignoreAttachments && data.Image != nil {
+			attachmentPath := path.Join("data", *data.Image)
+
+			zfile, ok := v.attachments[attachmentPath]
+			if !ok {
+				helpful := ""
+				candidates := v.findFileNameSuffix(*data.Image)
+				if len(candidates) != 0 {
+					helpful = "; we found a match outside the \"data/\" folder \"" + strings.Join(candidates, "\" or \"") + "\""
+				}
+
+				return &ImportValidationError{
+					ImportFileInfo: info,
+					FieldName:      "emoji.image",
+					Err:            fmt.Errorf("missing image file for emoji %s: %q%s", *data.Name, attachmentPath, helpful),
+				}
+			}
+
+			return v.validateSupportedImage(info, zfile)
+		}
+
+		return nil
+	})
+	if ive != nil {
+		return v.onError(ive)
+	}
+
+	return nil
+}
+
+func (v *Validator) Attachments() []string {
+	var used []string
+	for attachment := range v.attachmentsUsed {
+		used = append(used, attachment)
+	}
+	sort.Strings(used)
+	return used
+}
+
+func (v *Validator) UnusedAttachments() []string {
+	var unused []string
+	for attachment := range v.attachments {
+		if _, ok := v.attachmentsUsed[attachment]; !ok {
+			unused = append(unused, attachment)
+		}
+	}
+	sort.Strings(unused)
+	return unused
+}
+
+func (v *Validator) validateSupportedImage(info ImportFileInfo, zfile *zip.File) *ImportValidationError {
+	f, err := zfile.Open()
+	if err != nil {
+		return &ImportValidationError{
+			ImportFileInfo: info,
+			FieldName:      "emoji.image",
+			Err:            fmt.Errorf("error opening emoji image: %w", err),
+		}
+	}
+	defer f.Close()
+
+	if mime.TypeByExtension(strings.ToLower(path.Ext(zfile.Name))) == "image/svg+xml" {
+		var svg struct{}
+		err = xml.NewDecoder(f).Decode(&svg)
+		if err != nil {
+			return &ImportValidationError{
+				ImportFileInfo: info,
+				FieldName:      "emoji.image",
+				Err:            fmt.Errorf("error decoding emoji SVG file: %w", err),
+			}
+		}
+
+		return nil
+	}
+
+	_, _, err = image.Decode(f)
```

Comment:
For later, we might want to check the image dimensions first (`image.DecodeConfig`) and then decode, to avoid OOMs. We'll need to make a server API call to know the max allowed image dimensions. Probably a good HW item for contributors.

Reply:
Yeah, that's a good issue for HW.